docs/testing.rst

```diff
@@ -0,0 +1,155 @@
+
+    @pytest.fixture
+    @pytest.mark.freeze_time('2018-05-04')
+    def loggedin_app(app):
+        """Creates a logged-in test client instance."""
+        with app.test_client() as client:
+            with client.session_transaction() as sess:
+                sess['domain'] = 'example.com'
+                sess['google_oauth_token'] = {
+                    'access_token': 'this is totally fake',
+                    'id_token': 'this is not a real token',
+                    'token_type': 'Bearer',
+                    'expires_in': '3600',
+                    'expires_at': time.time() + 3600,
+                }
+            yield client
+
+Now that we have a logged-in client you can call any of the routes using the
+test client and check their responses.
+
+.. code-block:: python
+
+    def test_not_logged_in(app):
+        """Test that we redirect to Google to login."""
+        res = app.test_client().get('/')
+        assert ('redirected automatically to target URL: <a href="/login/'
+                'google">/login/google</a>').lower() in res.get_data(as_text=True).lower()
+        assert res.status_code == 302
+
+
+    def test_logged_in_index(loggedin_app):
+        """Tests getting the index route.
+
+        This will render the normal template as we have a valid oauth2 session.
+        """
+        res = loggedin_app.get('/')
+        assert res.content_type == 'text/html; charset=utf-8'
+        assert res.status_code == 200
+        assert 'something only shown when logged in' in res.get_data(as_text=True).lower()
+
+
+Calling authenticated APIs
+--------------------------
+
+Though we've managed to create a working session a problem now arises if you
+try to actually call an API, by using `google.get('some url')` for example.
+Your token will fail to validate and the request will be denied.
+
+This can be handled by a Python library called `responses`_, which lets us control
+the full HTTP request cycle.
+
+.. warning::
+    Note that this means we're essentially mocking the API we're calling, so
+    your tests will continue passing even if the real API has changed behaviour.
+
+Let's assume the index route calls out to the `Google Plus API`_ and displays some
+profile information. Here's how you could handle that.
+
+.. code-block:: python
+
+    import pytest
+    import responses
+
+
+    @responses.activate
+    def test_getting_profile(loggedin_app):
+        """Test displaying profile information."""
+        responses.add(
+            responses.GET,
+            'https://www.googleapis.com/plus/v1/people/me',
+            status=200,
+            json={
+              'kind': 'plus#person',
+              'id': '118051310819094153327',
+              'displayName': 'Chirag Shah',
+              'url': 'https://plus.google.com/118051310819094153327',
+              'image': {
+                'url': 'https://lh5.googleusercontent.com/-XnZDEoiF09Y/AAAAAAAAAAI/AAAAAAAAYCI/7fow4a2UTMU/photo.jpg'
```

Comment:
I just want to confirm that you're OK with having this content available in the public documentation for Flask-Dance. Some people might consider a name and photograph to be private information. Is this OK with you?

Reply:
Ah. This is a straight up copy from the Google Plus documentation: https://developers.google.com/+/web/api/rest/#data-formats. So from my p.o.v it's fine. I'm assuming it's some test data they purposefully created to have a more usable example.

Reply:
Ah, if Google has decided it's OK, then I'm fine with that. 👍